File: Project Euler/problem_35.cpp
// #include <iostream>
#include <bits/stdc++.h> 

typedef unsigned long long ull;
typedef unsigned long ul;

using namespace std;

bool isPrime(int n) {
    if (n < 2) return false;
    if (n == 2 || n == 3) return true;

    for (int i = 2; i < (int)sqrt(n) + 1; i++) {
        if (n % i == 0) return false;
    }

    return true;
}

vector<int> rotations(int n) {
    vector<int> res;
    string s = to_string(n);
    string rotation = *(s.end() - 1) + s.substr(0, s.size() - 1);
    while (rotation != s) {
        int x;
        stringstream ss;
        ss << rotation;
        ss >> x;
        res.push_back(x);
        ss.clear();
        rotation = *(rotation.end() - 1) + rotation.substr(0, rotation.size() - 1);
    }

    return res;
}

bool isCircular(int n, unordered_set<int> &primes) {
    for(auto p : rotations(n)) {
        if (primes.find(p) == primes.end()) return false;
    }
    // cout << n << endl;
    return true;
}

int main(int argc, const char** argv) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    unordered_set<int> primes;
    for (int i = 2; i < 1000000; i++) {
        if (isPrime(i)) primes.insert(i);
    }

    int count = 0;
    for(auto p : primes) {
        if (isCircular(p, primes)) count++;
    }

    cout << count << endl;
}

File: Project Euler/problem_35.py
#!/usr/bin/python3.6
from math import sqrt

def is_prime(n):
    if n < 2:
        return False
    if n == 2 or n == 3:
        return True
    
    for i in range(2, int(sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

def rotations(n):
    yield n
    n = str(n)
    rotation = n[-1] + n[:-1]
    while rotation != n:
        yield int(rotation)
        rotation = rotation[-1] + rotation[:-1]

def is_circular(n, primes):
    rots = rotations(n)
    res = True
    for p in rots:
        if p not in primes:
            res = False
            break
    return res

def count_circular_primes(maximum):
    primes = set()
    count = 0
    for i in range(2, 1000001):
        if is_prime(i):
            primes.add(i)
    for p in primes:
        if p > maximum:
            continue
        if is_circular(p, primes):
            count += 1
    return count
print(count_circular_primes(1000000))
# print(list(rotations(197)))

